porch/controllers/packagevariants/pkg/controllers/packagevariant/injection.go

```diff
@@ -0,0 +1,425 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package packagevariant
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sort"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	kptfilev1 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
+	porchapi "github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	api "github.com/GoogleContainerTools/kpt/porch/controllers/packagevariants/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ConfigInjectionAnnotation  = "kpt.dev/config-injection"
+	InjectedResourceAnnotation = "kpt.dev/injected-resource"
+)
+
+func injectionConditionType(object *fn.KubeObject) string {
+	return strings.Join([]string{"config", "injection", object.GetKind(), object.GetName()}, ".")
+}
+
+type injectableField struct {
+	Group string
+	Kind  string
+	Field string
+}
+
+// TODO: consider giving the admin control over this
+var allowedInjectionFields = []injectableField{
+	{Group: "", Kind: "ConfigMap", Field: "data"},
+	{Group: "*", Kind: "*", Field: "spec"},
+}
+
+type injectionPoint struct {
+	file               string
+	object             *fn.KubeObject
+	conditionType      string
+	required           bool
+	errors             []string
+	injected           bool
+	injectedName       string
+	inClusterResources []*unstructured.Unstructured
+}
+
+func newInjectionPoint(file string, object *fn.KubeObject) *injectionPoint {
+	annotations := object.GetAnnotations()
+	annotation, ok := annotations[ConfigInjectionAnnotation]
+	if !ok {
+		return nil
+	}
+	ip := &injectionPoint{
+		file:          file,
+		object:        object,
+		conditionType: injectionConditionType(object),
+	}
+	if annotation == "required" {
+		ip.required = true
+	} else if annotation == "optional" {
+		ip.required = false
+	} else {
+		ip.errors = append(ip.errors, fmt.Sprintf("%s: %s/%s has invalid %q annotation value of %q",
+			file, object.GetKind(), object.GetName(), ConfigInjectionAnnotation, annotation))
+	}
+	return ip
+}
+
+func ensureConfigInjection(ctx context.Context,
+	c client.Client,
+	pv *api.PackageVariant,
+	prr *porchapi.PackageRevisionResources) error {
+
+	if prr.Spec.Resources == nil {
+		return fmt.Errorf("nil resources found for PackageRevisionResources '%s/%s'", prr.Namespace, prr.Name)
```

Comment:
For this, in the interest of failing earlier rather than later, could we move this check out to `packagevariant_controller.go` right after we fetch the PackageRevisionResources (L753) , rather than waiting until we get all the way here to throw an error? 

fwiw, I see that `hashFromPackageRevisionResources` also uses `prr.Spec.Resources`

Reply:
Sure, good idea.